torch_np/_detail/implementations.py

```diff
@@ -19,11 +21,125 @@ def tensor_equal(a1_t, a2_t, equal_nan=False):
     return bool(result.all())
 
 
+def tensor_equiv(a1_t, a2_t):
+    # *almost* the same as tensor_equal: _equiv tries to broadcast, _equal does not
+    try:
+        a1_t, a2_t = torch.broadcast_tensors(a1_t, a2_t)
+    except RuntimeError:
+        # failed to broadcast => not equivalent
+        return False
+    return tensor_equal(a1_t, a2_t)
+
+
+def tensor_isclose(a, b, rtol=1.0e-5, atol=1.0e-8, equal_nan=False):
+    dtype = _dtypes_impl.result_type_impl((a.dtype, b.dtype))
+    a = a.to(dtype)
+    b = b.to(dtype)
+    result = torch.isclose(a, b, rtol=rtol, atol=atol, equal_nan=equal_nan)
+    return result
+
+
+# ### is arg real or complex valued ###
+
+
+def tensor_iscomplex(x):
+    if torch.is_complex(x):
+        return torch.as_tensor(x).imag != 0
+    result = torch.zeros_like(x, dtype=torch.bool)
+    if result.ndim == 0:
+        result = result.item()
+    return result
+
+
+def tensor_isreal(x):
+    if torch.is_complex(x):
+        return torch.as_tensor(x).imag == 0
+    result = torch.zeros_like(x, dtype=torch.bool)
+    if result.ndim == 0:
+        result = result.item()
+    return result
+
+
+def tensor_real_if_close(x, tol=100):
+    if not torch.is_complex(x):
+        return x
+    mask = torch.abs(x.imag) < tol * torch.finfo(x.dtype).eps
+    if mask.all():
+        return x.real
```

Comment:
This returns a view into the original tensor. We don't care too much about views / copies yet, but just FYI

Reply:
Does it?

```
In [2]: t = torch.arange(9)

In [3]: t.real._base

In [4]: t.real._base is None
Out[4]: True
```

Seems to be the same for the bare pytorch, numpy and the current wrapper package. Am I missing something?

Reply:
It's  not that it's a copy in the real case, but it's the tensor itself.
```python
>>> x = torch.arange(9)
>>> x.real is x
True
```

Reply:
Note that in the complex case it's a view of every other element in the tensor:
```python
>>> t = torch.arange(9).cfloat()
>>> t.real.stride()
(2,)
```

Reply:
Wow.
I suspect a program which critically depends on these things is an entry for a code golf competition, so I agree it's premature to seriously worry about it.

Reply:
Hmmmm, not really. If you are interested in performance and you don't care about autograd (for example, when you have deployed a trained model and you just want to run it), one of the things you want to minimise is the number of allocated tensors throughout the program. For that, views and writing-through-views are essential. If in one place you return a view, and in other you return a copy... then that's game.

Reply:
FTR, my background is _deeply_ caring for performance, and I'm regularly serving as a PITA reviewer with perf concerns in several other projects. And I do appreciate why views are important.

So _critically depends_ comment was about https://github.com/Quansight-Labs/numpy_pytorch_interop/pull/52#discussion_r1110032280 : where it's a copy vs view depending on the input dtype. 

In fact, I'd argue that if the view semantics is important here, the behavior of `torch.arange(9).real` is a bug.

Reply:
Note that `.real` is **never** a copy! It's either the original tensor or a view into the original tensor. A purist could say "oh, but you then should return the identity view onto the original tensor if the dtype is real" but it doesn't really matter. In this case, I do agree that if a program depends on being a view or the original tensor, then, well... you sort of deserve it.

Note that the PrimTorch project **does** go the extra mile and gets this one right. There is a `prims.view_of` function for this exact purpose. Again, I believe it's not necessary for us to do this.

Reply:
OK, am rolling back gratuitous `.clone()` calls, and let's keep this in mind for later.